controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java

```diff
@@ -71,111 +71,126 @@ public static Segment getSegment(final int number, final byte[] segmentTable) {
+            if (record.getEpoch() < truncationRecord.getTruncationEpochLow()) {
+                segments = Lists.newArrayList(truncationRecord.getStreamCut().keySet());
+            } else if (record.getEpoch() > truncationRecord.getTruncationEpochHigh()) {
+                // case 2: if record.epoch is after truncation, simply use the record epoch
                 segments = record.getSegments();
             } else {
-                // case 1: if record.epoch is before truncation, simply pick the truncation stream cut
-                if (record.getEpoch() < truncationRecord.getTruncationEpochLow()) {
-                    segments = Lists.newArrayList(truncationRecord.getStreamCut().keySet());
-                } else if (record.getEpoch() > truncationRecord.getTruncationEpochHigh()) {
-                    // case 2: if record.epoch is after truncation, simply use the record epoch
-                    segments = record.getSegments();
-                } else {
-                    // case 3: overlap between requested epoch and stream cut.
-                    // take segments from stream cut that are from or aftergit re this epoch.
-                    // take remaining segments from this epoch.
-                    segments = new ArrayList<>();
-                    // all segments from stream cut that have epoch >= this epoch
-                    List<Integer> fromStreamCut = truncationRecord.getCutEpochMap().entrySet().stream()
-                            .filter(x -> x.getValue() >= record.getEpoch())
-                            .map(Map.Entry::getKey)
-                            .collect(Collectors.toList());
-
-                    segments.addAll(fromStreamCut);
-                    // put remaining segments as those that dont overlap with ones taken from streamCut.
-                    segments.addAll(record.getSegments().stream().filter(x -> fromStreamCut.stream().noneMatch(y ->
-                            getSegment(x, segmentTable).overlaps(getSegment(y, segmentTable))))
-                            .collect(Collectors.toList()));
-                }
+                // case 3: overlap between requested epoch and stream cut.
+                // take segments from stream cut that are from or aftergit re this epoch.
+                // take remaining segments from this epoch.
+                segments = new ArrayList<>();
+                // all segments from stream cut that have epoch >= this epoch
+                List<Integer> fromStreamCut = truncationRecord.getCutEpochMap().entrySet().stream()
+                        .filter(x -> x.getValue() >= record.getEpoch())
+                        .map(Map.Entry::getKey)
+                        .collect(Collectors.toList());
+
+                segments.addAll(fromStreamCut);
+                // put remaining segments as those that dont overlap with ones taken from streamCut.
+                segments.addAll(record.getSegments().stream().filter(x -> fromStreamCut.stream().noneMatch(y ->
```

Comment:
1. No need for a collector here. Simply invoke `.forEach(segments::add)` which will prevent the additional list creation.

2. Also, you are looping several times over `fromStreamCut` and invoking `getSegment(y, segmentIndex, segmentTable)`. I suggest you pre-cache these values in a local map.

Reply:
> Also, you are looping several times over fromStreamCut and invoking getSegment(y, segmentIndex, segmentTable). I suggest you pre-cache these values in a local map.

getSegment is an O(1) fetch from the segment and index blobs. Its not a remote call.